games/xworld3d/x3scene.cpp

```diff
@@ -0,0 +1,160 @@
+namespace xworld3d {
+
+using simulator::util::path_join;
+
+X3Scene::X3Scene(float gravity, float time_step, int frame_skip):
+        frame_skip_(frame_skip),
+        img_height_(FLAGS_x3_img_height),
+        img_width_(FLAGS_x3_img_width),
+        world_(gravity, time_step) {
+    dt_ = time_step * frame_skip_;
+    world_.set_glsl_path(FLAGS_x3_glsl_path);
+    camera_ = std::unique_ptr<X3Camera>(
+            new X3Camera(world_, img_height_, img_width_));
+    blocks_.clear();
+    agents_.clear();
+    goals_.clear();
+}
+
+void X3Scene::build_scene(const std::vector<X3ItemInfo>& infos,
+                          const std::string& floor_file,
+                          const std::string& stadium_file) {
+    clear_scene();
+
+    std::vector<Object> olist = world_.load_mjcf(floor_file);
+    for (auto& o : olist) {
+        o.query_position();
+    }
+
+    Pose stadium_pose(0.0f, 0.0f, -0.05f);
+    auto t = world_.load_thingy(stadium_file, stadium_pose,
+                                0.01f, 0.0f, 0xffffff, true);
+
+    for (auto const& i : infos) {
+        add_item(i);
+    }
+    CHECK_GT(agents_.size(), 0) << "There should be at least one agent.";
+}
+
+void X3Scene::add_item(const X3ItemInfo& info) {
+    if (info.type == AGENT) {
+        CHECK(agents_.find(info.name) == agents_.end())
+                << "An agent with name " << info.name << " exists.";
+        auto agent = std::make_shared<X3Agent>(info, world_,
+                                               FLAGS_x3_speed_norm,
+                                               FLAGS_x3_orientation_bins,
+                                               FLAGS_x3_reaching_distance);
+        agents_[info.name] = agent;
+    } else if (info.type == GOAL) {
+        CHECK(goals_.find(info.name) == goals_.end())
+                << "A goal with name " << info.name << " exists.";
+        auto goal = std::make_shared<X3Goal>(info, world_);
+        goals_[info.name] = goal;
+    } else if (info.type == BLOCK) {
+        CHECK(blocks_.find(info.name) == blocks_.end())
+                << "An item with name " << info.name << " exists.";
+        auto block = std::make_shared<X3Block>(info, world_);
+        blocks_[info.name] = block;
+    } else {
+        LOG(ERROR) << "unknown item type: " << info.type;
+    }
+}
+
+void X3Scene::apply_action(const X3AgentPtr& agent, int a,
+                           float& reward, int& event) {
+    reward = STEP_PENALTY;
+    event = X3Event::NOTHING;
+
+    switch (a) {
+        case X3NavAction::MOVE_FORWARD:
+            agent->move_forward();
+            break;
+        case X3NavAction::TURN_LEFT:
+            agent->turn_left();
+            break;
+        case X3NavAction::TURN_RIGHT:
+            agent->turn_right();
+            break;
+        case X3NavAction::JUMP:
+            agent->jump();
+            break;
+        case X3NavAction::COLLECT:
+            {
+                X3Goal* goal = collect_goal(agent);
```

Comment:
avoid using raw pointers

Reply:
we should use raw pointers here as we dont want to claim ownership here